// rzkyfdlh-api/lib/toresult.js
const toResult = async (json, options = {}) => {
if(!json) return new Error('Masukan result, example: {\nnama: `Rzky`,\n result: `ini result`\n}')
  let { arrow, head, upper, down, line } = {
    head: "âœ¦",
    arrow: "âœ§",
    body: "â”ƒ",
    upper: "â”",
    down: "â”—",
    line: "â”",
    wings: ["ğ“†© ", " ğ“†ª"],
    needed: ["<", ">"],
    optional: ["(", ")"],
  };
  let opts = {
    unicode: true,
    ignoreVal: [null, undefined],
    ignoreKey: [],
    title: "Parse Result",
    headers: `${head}${line.repeat(4)}${arrow} Â» %title Â«`,
    body: `âœ *%key*: _%value_`,
    footer: head + line + line + line + arrow + "\n",
    ...options,
  };
  let { unicode, ignoreKey, title, headers, ignoreVal, body, footer } = opts;

  let obj = Object.entries(json);
  let tmp = [];
  for (let [_key, val] of obj) {
    if (ignoreVal.indexOf(val) !== -1) continue;
    let key = _key[0].toUpperCase() + _key.slice(1);
    let type = typeof val;
    if (ignoreKey && ignoreKey.includes(_key)) continue;
    switch (type) {
      case "boolean":
        tmp.push([key, val ? "Ya" : "Tidak"]);
        break;
      case "object":
        if (Array.isArray(val)) {
          tmp.push([key, val.join(", ")]);
        } else {
          tmp.push([
            key,
            toResult(val, {
              ignoreKey,
              unicode: false,
            }),
          ]);
        }
        break;
      default:
        tmp.push([key, val]);
        break;
    }
  }
  if (unicode) {
    let text = [
      headers.replace(/%title/g, title),
      tmp
        .map((v) => {
          return body.replace(/%key/g, v[0]).replace(/%value/g, v[1]);
        })
        .join("\n"),
      footer,
    ];
    return text.join("\n").trim();
  }
  return tmp;
};

module.exports = { toResult }
